### PR TITLE
[WIP] Make compel work with SD-XL

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR allows to use the Compel library with SD-XL. This is a first proposal.

The reason I went for this design is because the different text encoders even use different tokenizers (they don't match 100% - the pad token ID is different).

The following seems to work (I have not tested it thoroughly):

```py
from diffusers import DiffusionPipeline
from compel import Compel
import torch

pipeline = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-0.9", variant="fp16", use_safetensors=True, torch_dtype=torch.float16).to("cuda")
compel = Compel(tokenizer=[pipeline.tokenizer, pipeline.tokenizer_2] , text_encoder=[pipeline.text_encoder, pipeline.text_encoder_2], use_penultimate_clip_layer=True, use_penultimate_layer_norm=False, requires_pooled=[False, True])

# ++ or -- "ball"

prompt = "a cat playing with a ball++ in the forest"
# prompt = "a cat playing with a ball-- in the forest"
conditioning, pooled = compel(prompt)

# generate image
image = pipeline(prompt_embeds=conditioning, pooled_prompt_embeds=pooled, num_inference_steps=30).images[0]
```

For `"ball++"` I'm gettin
ball_plus_plus

Fro `"ball--"` I'm getting:
ball_minus_minus

### CODE REVIEW

File: src/compel/embeddings_provider.py [hunk 1]
```diff
@@ -108,7 +114,7 @@ def get_embeddings_for_weighted_prompt_fragments(self,
 
             # handle weights >=1
             tokens, per_token_weights, mask = self.get_token_ids_and_expand_weights(fragments, weights, device=device)
-            base_embedding = self.build_weighted_embedding_tensor(tokens, per_token_weights, mask, device=device)
+            base_embedding = self.build_weighted_embedding_tensor(tokens, per_token_weights, mask)[0]
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:117]:
```suggestion
            base_embedding = self.build_weighted_embedding_tensor(tokens, per_token_weights, mask, device=device)
```

File: src/compel/embeddings_provider.py [hunk 2]
```diff
@@ -177,11 +183,15 @@ def get_embeddings_for_weighted_prompt_fragments(self,
...
         if should_return_tokens:
-            return batch_z, batch_tokens
-        else:
-            return batch_z
+            outputs += (batch_tokens,)
+
+        if len(outputs) == 1:
+            return outputs[0]
+
+        return outputs
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:194]:
```suggestion
        if should_return_tokens:
            return batch_z, batch_tokens
        else:
            return batch_z
```

File: src/compel/embeddings_provider.py [hunk 3]
```diff
@@ -177,11 +183,15 @@ def get_embeddings_for_weighted_prompt_fragments(self,
 
         # should have shape (B, 77, 768)
         #print(f"assembled all tokens into tensor of shape {batch_z.shape}")
+        outputs = (batch_z,)
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:186]:
```suggestion
```

File: src/compel/embeddings_provider.py [hunk 4]
```diff
@@ -178,6 +184,7 @@ def get_embeddings_for_weighted_prompt_fragments(self,
         # should have shape (B, 77, 768)
         #print(f"assembled all tokens into tensor of shape {batch_z.shape}")
 
+        if should_return_tokens:
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:187]:
```suggestion
```

File: src/compel/embeddings_provider.py [hunk 5]
```diff
@@ -302,10 +309,9 @@ def build_weighted_embedding_tensor(self,
                                         token_ids: torch.Tensor,
                                         per_token_weights: torch.Tensor,
                                         attention_mask: Optional[torch.Tensor] = None,
+                                        return_pooled: bool = False,
                                         device: Optional[str] = None) -> torch.Tensor:
         """
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:314]:
```suggestion
        """
        Build a tensor that embeds the passed-in token IDs and applies the given per_token weights
        
```

File: src/compel/embeddings_provider.py [hunk 6]
```diff
@@ -328,8 +337,9 @@ def build_weighted_embedding_tensor(self,
                                        [self.tokenizer.eos_token_id] +
                                        [self.tokenizer.pad_token_id] * (self.max_token_count - 2),
                                        dtype=torch.int, device=device).unsqueeze(0)
-        empty_z = self._encode_token_ids_to_embeddings(empty_token_ids)
+        empty_z, _ = self._encode_token_ids_to_embeddings(empty_token_ids)
         weighted_z = None
+        pooled = None
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:342]:
```suggestion
```

File: src/compel/embeddings_provider.py [hunk 7]
```diff
@@ -302,10 +312,9 @@ def build_weighted_embedding_tensor(self,
                                         token_ids: torch.Tensor,
                                         per_token_weights: torch.Tensor,
                                         attention_mask: Optional[torch.Tensor] = None,
+                                        return_pooled: bool = False,
                                         device: Optional[str] = None) -> torch.Tensor:
         """
-        Build a tensor that embeds the passed-in token IDs and applies the given per_token weights
-
         :param token_ids: A tensor of shape `n*[self.max_length]` containing token IDs (ints) where n is some arbitrary
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:318]:
```suggestion
        Build a tensor that embeds the passed-in token IDs and applies the given per_token weights
        
        :param token_ids: A tensor of shape `n*[self.max_length]` containing token IDs (ints) where n is some arbitrary
```

File: src/compel/embeddings_provider.py (hunk 2 above)
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:194]:
```suggestion
            return batch_z, batch_tokens
        else:
            return batch_z
```

File: src/compel/embeddings_provider.py [hunk 8]
```diff
@@ -328,7 +336,7 @@ def build_weighted_embedding_tensor(self,
                                        [self.tokenizer.eos_token_id] +
                                        [self.tokenizer.pad_token_id] * (self.max_token_count - 2),
                                        dtype=torch.int, device=device).unsqueeze(0)
-        empty_z = self._encode_token_ids_to_embeddings(empty_token_ids)
+        empty_z, _ = self._encode_token_ids_to_embeddings(empty_token_ids)
         weighted_z = None
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:340]:
```suggestion
        weighted_z = None
        pooled = None
```

File: src/compel/compel.py [hunk 9]
```diff
@@ -170,8 +189,6 @@ def build_conditioning_tensor_for_prompt_object(self, prompt: Union[Blend, Flatt
 
         raise ValueError(f"unsupported prompt type: {type(prompt).__name__}")
 
-
-
     def pad_conditioning_tensors_to_same_length(self, conditionings: List[torch.Tensor],
```
Comment (patrickvonplaten) [inline src/compel/compel.py:192]:
```suggestion


    def pad_conditioning_tensors_to_same_length(self, conditionings: List[torch.Tensor],
```

File: src/compel/compel.py [hunk 10]
```diff
@@ -222,10 +239,16 @@ def _get_conditioning_for_flattened_prompt(self,
         conditioning, tokens = self.conditioning_provider.get_embeddings_for_weighted_prompt_fragments(
             text_batch=[fragments], fragment_weights_batch=[weights],
             should_return_tokens=True, device=self.device)
+
+        outputs = (conditioning,)
+
```
Comment (patrickvonplaten) [inline src/compel/compel.py:244]:
```suggestion
```

File: src/compel/compel.py [hunk 11]
```diff
@@ -222,10 +239,16 @@ def _get_conditioning_for_flattened_prompt(self,
...
         if should_return_tokens:
-            return conditioning, tokens
-        else:
-            return conditioning
+            outputs += (tokens,)
+
+        if len(outputs) == 1:
+            return outputs[0]
+
+        return outputs
```
Comment (patrickvonplaten) [inline src/compel/compel.py:251]:
```suggestion
            return conditioning, tokens
        else:
            return conditioning
```

File: src/compel/embeddings_provider.py [hunk 12]
```diff
@@ -94,6 +97,7 @@ def get_embeddings_for_weighted_prompt_fragments(self,
 
         batch_z = None
         batch_tokens = None
+
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:100]:
```suggestion
```

File: src/compel/embeddings_provider.py [hunk 13]
```diff
@@ -304,8 +321,6 @@ def build_weighted_embedding_tensor(self,
                                         attention_mask: Optional[torch.Tensor] = None,
                                         device: Optional[str] = None) -> torch.Tensor:
         """
-        Build a tensor that embeds the passed-in token IDs and applies the given per_token weights
-
         :param token_ids: A tensor of shape `n*[self.max_length]` containing token IDs (ints) where n is some arbitrary
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:324]:
```suggestion
        Build a tensor that embeds the passed-in token IDs and applies the given per_token weights
        
        :param token_ids: A tensor of shape `n*[self.max_length]` containing token IDs (ints) where n is some arbitrary
```

File: src/compel/embeddings_provider.py [hunk 14]
```diff
@@ -177,13 +180,17 @@ def get_embeddings_for_weighted_prompt_fragments(self,
 
         # should have shape (B, 77, 768)
         #print(f"assembled all tokens into tensor of shape {batch_z.shape}")
+        outputs = (batch_z,)
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:183]:
```suggestion
```

File: src/compel/embeddings_provider.py [hunk 15]
```diff
@@ -179,11 +182,14 @@ def get_embeddings_for_weighted_prompt_fragments(self,
...
         if should_return_tokens:
-            return batch_z, batch_tokens
-        else:
-            return batch_z
+            outputs += (batch_tokens,)
+
+        if len(outputs) == 1:
+            return outputs[0]
+
+        return outputs
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:190]:
```suggestion
            return batch_z, batch_tokens
        else:
            return batch_z
```

File: src/compel/embeddings_provider.py [hunk 16]
```diff
@@ -363,10 +377,14 @@ def _encode_token_ids_to_embeddings(self, token_ids: torch.Tensor,
                                                 attention_mask,
                                                 output_hidden_states=self.use_penultimate_clip_layer,
                                                 return_dict=True)
+
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:380]:
```suggestion
```

File: test/prompting_test_utils.py [hunk 17]
```diff
@@ -75,15 +79,14 @@ def hidden_states(self):
...
 
     def __call__(self, input_ids, attention_mask=None, **kwargs):
-        return self.forward(input_ids=input_ids, attention_mask=attention_mask, return_dict=True)
+        return self.forward(input_ids=input_ids, attention_mask=attention_mask, return_dict=True, output_hidden_states=kwargs.pop("output_hidden_states", False))
 
     @property
     def text_model(self):
         tm = Mock()
         tm.final_layer_norm = nn.LayerNorm(normalized_shape=[self.text_model_max_length, self.embedding_length])
         return tm
```
Comment (patrickvonplaten) [inline test/prompting_test_utils.py:89]:
```suggestion


```

File: src/compel/embeddings_provider.py [hunk 18]
```diff
@@ -220,6 +223,16 @@ def get_token_ids(self, texts: List[str], include_start_and_end_markers: bool =
 
         return result
 
+    def maybe_get_pooled(self, texts: List[str], attention_mask: Optional[torch.Tensor]=None) -> Optional[torch.Tensor]:
+        if not self.requires_pooled:
+            return None
+
```
Comment (patrickvonplaten) [inline src/compel/embeddings_provider.py:229]:
```suggestion
```

File: src/compel/compel.py [hunk 19]
```diff
@@ -3,11 +3,11 @@
 
 import torch
 from torch import Tensor
-from transformers import CLIPTokenizer, CLIPTextModel
+from transformers import CLIPTokenizer, CLIPTextModel, ConditionalDetrImageProcessor
```
Comment (patrickvonplaten) [inline src/compel/compel.py:6]:
```suggestion
from transformers import CLIPTokenizer, CLIPTextModel
```

File: src/compel/compel.py [hunk 20]
```diff
@@ -50,22 +53,42 @@ def __init__(self,
...
+                                                            )
+
         self._device = device
 
     @property
     def device(self):
         return self._device if self._device else self.conditioning_provider.text_encoder.device
 
-    def make_conditioning_scheduler(self, positive_prompt: str, negative_prompt: str='') -> ConditioningScheduler:
+    def make_conditioning_scheduler(self, positive_prompt: str, negative_prompt: str=''):
```
Comment (patrickvonplaten) [inline src/compel/compel.py:91]:
```suggestion
    def make_conditioning_scheduler(self, positive_prompt: str, negative_prompt: str='')  -> ConditioningScheduler:
```

File: test/prompting_test_utils.py [hunk 21]
```diff
@@ -64,18 +64,26 @@ def __init__(self, last_hidden_state):
...
 
             def __getitem__(self, item):
-                assert item == 0
-                return self.last_hidden_state
+                if item == 0:
+                    return self.last_hidden_state[:, -1, :]
+                if item == 1:
+                    return self.last_hidden_state
+                if item == 2:
+                    return 2 * [self.last_hidden_state]
```
Comment (patrickvonplaten) [inline test/prompting_test_utils.py:72]:
```suggestion
                assert item == 0
                return self.last_hidden_state
```

**Comment by damian0815** [comment]
@patrickvonplaten i had a closer look and there's a couple of issues. first of all this simply won't work:

`pooled = self.conditioning_provider.maybe_get_pooled([text])`

this will passing all of the Compel syntax markers into CLIP, which is not good.

instead i think this should be designed as follows:

1. on EmbeddingsProvider, replace the `use_penultimate_clip_layer`, `use_penultimate_layer_norm` and `requires_pooled` with an enum `ReturnedEmbeddingsType` which has options `PooledOutputs`, `PenultimateHiddenStates`, `PenultimateHiddenStatesNormalized` and `LastHiddenStatesNormalized`.
2. for SDXL support, make 3 EmbeddingsProviders, not just 2 - one for the Refiner text encoder, and two for the Base text encoder, one passing `PenultimateHiddenStatesNormalized` and the other passing `PooledOutputs`. this introduces a performance penalty because we'll be hitting the text encoder twice when we could be re-using the outputs, but i think it's worth it to avoid turning the logic into spaghetti

**Comment by patrickvonplaten** [comment]
1.) For the pooled embedding vector I don't think it's a problem that all syntax markers are passed as it's a pooled vector not a sequential hidden states vector. E.g. for an input sentence:

```
A cat AND a ball
```

where `AND` is a syntax marker IMO it's better to do:

```py
pooled = pooled_vector_of("A cat AND a ball")
```

compared to:

```py
pooled = torch.mean([pooled_vector_of("A cat"), pooled_vector_of("A ball")]
```

or some other "merging" operation.

Pooling vectors only need to get the "general" gist of the input text, they don't need to have contextualized embeddings for every token in the input.

The way I understood it: `"A cat AND a ball"` will be split into multiple forward passes for the clip cross attention embedding, which is not what we want to do for pooling IMO. For clip cross attention this is fine / good because here we rather do torch.concat(...) [here](https://github.com/damian0815/compel/blob/fe59a4665263b949d685eb33aa3871b05d51a0c6/src/compel/embeddings_provider.py#L152) which is not "distribution-destructive" compared to merged two pooled vectors IMO. 

What do you think? 

2.) Regrading sequences that run over the max limit I'd just truncate / cut them for now. We could revisit later if we think it makes sense, but the orginal implementation also cuts it for now (see [here](https://github.com/Stability-AI/generative-models/blob/5c10deee76adad0032b412294130090932317a87/sgm/modules/encoders/modules.py#L457)).

RE: 

> instead i think this should be designed as follows:
> 
> on EmbeddingsProvider, replace the use_penultimate_clip_layer, use_penultimate_layer_norm and requires_pooled arguments to __init__ with an enum ReturnedEmbeddingsType which has options PooledOutputs, PenultimateHiddenStatesNonNormalized, PenultimateHiddenStatesNormalized and LastHiddenStatesNormalized.
> for SDXL support, make 3 EmbeddingsProviders, not just 2 - one for the Refiner text encoder passing i guess PenultimateHiddenStatesNormalized, and two for the Base text encoder, one passing PenultimateHiddenStatesNonNormalized and the other passing PooledOutputs (i may have got those wrong but you get the idea i hope). this introduces a performance penalty because we'll be hitting the text encoders three times instead of only twice when we could be re-using the outputs, but i think it's worth it to avoid turning the logic into spaghetti.
> doing number 2. at least will solve the problem of having to make a separate call to get the pooled outputs, and you'll get syntax handling on the pooled outputs the same as for the non-pooled outputs.

I don't fully understand. Do you maybe just want to take over the PR? Feel free to open a new one if this one is too far away from what you were thinking.

**Comment by bghira** [comment]
> 2.) Regrading sequences that run over the max limit I'd just truncate / cut them for now. We could revisit later if we think it makes sense, but the orginal implementation also cuts it for now (see [here](https://github.com/Stability-AI/generative-models/blob/5c10deee76adad0032b412294130090932317a87/sgm/modules/encoders/modules.py#L457)).

the original implementation also uses Euler and doesn't do half of the neat tricks that Compel can do, such as textual inversions or LoRA parsing. I don't think it should be used as the goal for comparison.

one of the main reasons to use Compel has been for long prompt weighting. the effect of the attention emphasis in SDXL is already weaker than in previous models, so, I don't see any purpose of using Compel with SDXL if there's no long prompt parsing. thoughts?

**Comment by patrickvonplaten** [comment]
Note that we're just talking about "long prompt parsing" of the pooled vector where it's much less clear how we can average multiple pooled vectors. The cross attention vectors are just like before "long prompt parsed". We cannot provide a pooled prompt vector of arbitrary length by definition (contrary to the cross attention vectors).

**Comment by bghira** [comment]
ohhhh. i am sorry for the confusion. that is my unfamiliarity with this.

**Pull request merged**